app/src/main/java/com/kickstarter/viewmodels/DiscoveryFragmentViewModel.kt

```diff
@@ -0,0 +1,583 @@
+
+            val projectOnUserChangeSave = loggedInUserOnHeartClick
+                .switchMap {
+                    this.toggleProjectSave(it.second)
+                }
+                .share()
+
+            loggedOutUserOnHeartClick
+                .map { it }
+                .subscribe {
+                    this.startLoginToutActivityToSaveProject.onNext(it.second)
+                }
+
+            val savedProjectOnLoginSuccess = this.startLoginToutActivityToSaveProject
+                .compose(Transformers.combineLatestPair(this.currentUser.observable()))
+                .filter { su ->
+                    su.second != null
+                }.take(1)
+                .switchMap {
+                    this.saveProject(it.first)
+                }
+                .share()
+
+            this.projectList
+                .compose(Transformers.takePairWhen(projectOnUserChangeSave))
+                .map {
+                    it.second.updateStartedProjectAndDiscoveryParamsList(it.first)
+                }
+                .distinctUntilChanged()
+                .compose(bindToLifecycle())
+                .subscribe {
+                    this.projectList.onNext(it)
+                }
+
+            this.projectList
+                .compose(Transformers.takePairWhen(savedProjectOnLoginSuccess))
+                .map {
+                    it.first.indexOfFirst { item ->
+                        item.first.id() == it.second.id() && item.first.slug() == it.second.slug()
+                    }
+                }
+                .distinctUntilChanged()
+                .delay(300, TimeUnit.MILLISECONDS, environment.scheduler())
+                .compose(bindToLifecycle())
+                .subscribe {
+                    scrollToSavedProjectPosition.onNext(it)
+                }
+
+            val projectSavedStatus = projectOnUserChangeSave.mergeWith(savedProjectOnLoginSuccess)
+
+            projectSavedStatus
+                .compose(bindToLifecycle())
+                .subscribe { this.analyticEvents.trackWatchProjectCTA(it, DISCOVER) }
+
+            projectSavedStatus
+                .filter { p -> p.isStarred() && p.isLive && !p.isApproachingDeadline }
+                .compose(Transformers.ignoreValues())
+                .compose(bindToLifecycle())
+                .subscribe(this.showSavedPrompt)
+        }
+
+        /**
+         * Calls to GraphQL client to fetch projects filtering by DiscoveryParams
+         * @param discoveryParamsStringPair .first discovery params.
+         * @param discoveryParamsStringPair .second cursor for pagination, null on the first call.
+         * @return Observable<DiscoverEnvelope>
+         </DiscoverEnvelope> */
+        private fun makeCallWithParams(discoveryParamsStringPair: Pair<DiscoveryParams?, String?>): Observable<DiscoverEnvelope> {
+            return apolloClient.getProjects(
+                discoveryParamsStringPair.first!!,
```

Comment:
Is it okay that we're forcing the first item in the pair open without checking if it's there?

Reply:
Done☺️